File: FuncionsPP.py
import skimage as ski
import numpy as np
import glob

def generarMatrius():
    archivos = glob.glob("img/*.jpg")
    primera_img = ski.io.imread(archivos[0], as_gray=True)
    alto, ancho = primera_img.shape

    num_fotos = len(archivos)
    matfotos = np.zeros((num_fotos, alto, ancho), dtype=np.uint8)

    for i in range(num_fotos):
        img = ski.io.imread(archivos[i], as_gray=True)
        matfotos[i] = (img * 255).astype(np.uint8)
            
    return matfotos

def calcularModelFons(train_mat):
    mitjana = np.mean(train_mat, axis=0)
    desviacio = np.std(train_mat, axis=0)
    return mitjana, desviacio

def tasca_3(llindar, mitjana, img):
    resta = np.abs(img - mitjana)
    deteccio_binaria = (resta > llindar) * 255
    return deteccio_binaria

def tasca_4(alpha, beta, mitjana, desviacio, img):
    diff = np.abs(img - mitjana)
    llindar_adaptatiu = alpha * desviacio + beta
    mascara = (diff > llindar_adaptatiu) * 255
    return mascara

def generar_tots_resultats(opcio, test_mat, llindar, alpha, beta, mitjana, desviacio):
    results_seq = []
    for i in range(len(test_mat)):

        if(opcio == 3):
            mask = tasca_3(llindar, mitjana, test_mat[i])
        else:
            mask = tasca_4(alpha, beta, mitjana, desviacio, test_mat[i])
        results_seq.append(mask)
    results_seq = np.array(results_seq)
    return results_seq

#Funcion per carregar les imatges de test i les groundtruths
def carregar_groundtruth_test():
    
    gt_files = (glob.glob("groundtruth/*.png"))
    
    gt_test_files = []
    
    for f in gt_files:
        num = int(f.split("gt")[1].split(".")[0])
        if 1201 <= num <= 1350:
            gt_test_files.append(f)
    
    gt_test_files = sorted(gt_test_files)
    
    primera = io.imread(gt_test_files[0], as_gray=True)
    h, w = primera.shape
    
    gt_mat = np.zeros((len(gt_test_files), h, w), dtype=np.uint8)
    
    for i in range(len(gt_test_files)):
        gt_mat[i] = io.imread(gt_test_files[i], as_gray=True)
        
    return gt_mat

def calcular_accuracy(pred_seq, gt_seq):
    
    accuracies = []
    
    for i in range(len(pred_seq)):
        
        pred_bin = (pred_seq[i] == 255)
        gt_bin = (gt_seq[i] == 255)
        
        accuracy = np.mean(pred_bin == gt_bin)
        accuracies.append(accuracy)
        
    accuracy_mitjana = np.mean(accuracies)
    
    return accuracy_mitjana, accuracies

File: FuncionsPP.m
classdef FuncionsPP
    methods (Static)
        function matfotos = generarMatrius()
            archivos = dir('img/*.jpg');
            num_fotos = length(archivos);
            
            % llegir primera img per saber el tamany
            primera_img = imread(fullfile(archivos(1).folder, archivos(1).name));
            if size(primera_img, 3) == 3
                primera_img = rgb2gray(primera_img);
            end
            [alto, ancho] = size(primera_img);
            
            % crear matriu (H x W x Num)
            matfotos = zeros(alto, ancho, num_fotos, 'uint8');
            
            for i = 1:num_fotos
                img = imread(fullfile(archivos(i).folder, archivos(i).name));
                if size(img, 3) == 3
                    img = rgb2gray(img);
                end
                matfotos(:,:,i) = img;
            end
        end

        function [mitjana, desviacio] = calcularModelFons(train_mat)
            train_double = double(train_mat);
            mitjana = mean(train_double, 3);
            desviacio = std(train_double, 0, 3);
        end

        function mascara = tasca_3(llindar, mitjana, img)
            resta = abs(double(img) - mitjana);
            mascara = resta > llindar; 
            mascara = uint8(mascara * 255); 
        end

        function mascara = tasca_4(alpha, beta, mitjana, desviacio, img)
            diff = abs(double(img) - mitjana);
            llindar_adaptatiu = alpha * desviacio + beta;
            mascara = uint8((diff > llindar_adaptatiu) * 255);
        end

        function res = generar_tots_resultats(opcio, test_mat, llindar, alpha, beta, mitjana, desviacio)
            [H, W, num_frames] = size(test_mat);
            res = zeros(H, W, num_frames, 'uint8');
            
            for i = 1:num_frames
                if opcio == 3
                    res(:,:,i) = FuncionsPP.tasca_3(llindar, mitjana, test_mat(:,:,i));
                else
                    res(:,:,i) = FuncionsPP.tasca_4(alpha, beta, mitjana, desviacio, test_mat(:,:,i));
                end
            end
        end

        function gt_mat = carregar_groundtruth_test()
            archivos = dir('groundtruth/*.png');
            count = 1;
            for i = 1:length(archivos)
                num = str2double(regexp(archivos(i).name, '\d+', 'match'));
                if num >= 1201 && num <= 1350
                    gt_files{count} = fullfile(archivos(i).folder, archivos(i).name);
                    count = count + 1;
                end
            end
            
            temp = imread(gt_files{1});
            [H, W] = size(temp);
            gt_mat = zeros(H, W, length(gt_files), 'uint8');
            
            for i = 1:length(gt_files)
                img = imread(gt_files{i});
                if size(img, 3) == 3, img = rgb2gray(img); end
                gt_mat(:,:,i) = img;
            end
        end

        function [accuracy_mitjana, acc_per_img] = calcular_accuracy(pred_seq, gt_seq)
            num_frames = size(pred_seq, 3);
            acc_per_img = zeros(num_frames, 1);
            
            for i = 1:num_frames
                pred_bin = pred_seq(:,:,i) == 255;
                gt_bin = gt_seq(:,:,i) == 255;
                
                acc_per_img(i) = sum(pred_bin(:) == gt_bin(:)) / numel(pred_bin);
            end
            accuracy_mitjana = mean(acc_per_img);
        end
    end
end